designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,556 @@
+
+
+## Multiple imputation for missing data
+
+One standard approach to multiple imputation is to use an imputation
+model on the data, then propgate multiple data sets through to
+inference.  That is, we start with a data set $y^\text{inc}$ with
+missing data, then impute $y^{\text{miss}(n)}$ for multiple $n$ and
+define a total of $N$ complete data sets $y^{(n)} = y^\text{inc},
+y^{\text{miss}(n)}$.  For each of these $N$ data sets, we take
+$M posterior draws, $\theta^{(n, m)} \sim p(\theta \mid y^{(n)})$.  We
+then perform standard plug-in Monte Carlo inference with all $N \cdot
+M$ draws.
+
+We will assume that there is a simple missing count data item and we
+will deal with the missingness through a Poisson regression.
+
+```stan
+data {
+  int<lower=0> N;
+  vector[N] x1_obs, x2_obs;
+  int<lower=0, upper=1>[N] miss1, miss2; 
+  vector[N] y;
+}
+parameters {
+  real alpha1, alpha2;
+  real<lower=0> sigma1, sigma2;
+  real beta1, beta2;
+  real<lower=0> sigma;
+  
+}
+latent data {
+  vector[N] x1 = x1_obs, x2 = x_obs;
+  for (n in 1:N) {
+    if (missing1[n]) {
+      x1[n] = normal_rng(alpha1 * x2[n], sigma1);
+    }
+    if (missing2[n]) {
+      x2[n] = normal_rng(alpha2 * x1[n], sigma2);
+    }
+  }
+}
+model {
+  x1 ~ normal(alpha1 * x2, sigma1); 
+  x2 ~ normal(alpha2 * x1, sigma2);
+  y ~ normal(beta1 * x1 + beta2 * x2, sigma);
+}
+```
+
+The result will be a multiple imputation, not a coherent joint
+Bayesian model.  This will work even if the missing data is discrete
+or the conditional distributions are not coherent. 
+
+## Stochastic gradient descent
+
+The usual model in Stan evaluates the likelihood for the given data
+set and a prior.  There has, until now, never been a good way to do
+stochastic gradient descent over subsampled data.  Now we can do
+that.  This is the simplest possible example.  Suppose we start with
+this model.
+
+```stan
+data {
+  int<lower=0> N;
+  array[N] int<lower=0, upper=1> y;
+}
+parameters {
+  real<lower=0, upper=1> theta;
+}
+model {
+  theta ~ beta(5, 5);
+  y ~ bernoulli(theta);
+}
+```
+
+We can move to stochastic gradient descent by randomly subsampling `y`
+in the `latent data` block.  As we do this, it's traditional to
+scale the likelihood to match the original data size.
+
+
+```stan
+data {
+  int<lower=0> N;
+  int<lower=0> N_sub;
+  array[N] int<lower=0, upper=1> y;
+}
+transformed data {
+  real one_over_N_sub = 1.0 / N_sub;
+  real N_over_N_sub = N * one_over_N_sub;
+  simplex[N_sub] unif = rep_vector(one_over_N_sub, N_sub);
+}
+parameters {
+  real<lower=0, upper=1> theta;
+}
+latent data {
+  array[N_sub] y_sub;
+  for (n in 1:N_sub) {
+    y_sub[n] = categorical_rng(unif);
+  }
+}
+model {
+  theta ~ beta(5, 5);
+  # scale data weight to match population size
+  target += N_over_N_sub * bernoulli_lpmf(y_sub | theta);
+}
+```
+
+If you plug this model into optimization, by default it will apply
+quasi-Newton steps using stochastic gradient.
```

Comment:
I'm not sure what reference you're looking for.

1.  That's right.  Most SGD implementations don't converge.  That's often considered an advantage in the ML world for its implicit regularization.

2.  It won't converge.  See 1.  That's right for L-BFGS with limited memory, but if you use BFGS it has a more global memory, but still only along a single trajectory.

Reply:
The algorithm proposed in this section is L-BFGS with stochastic data subsampling. My expectation is that it will not find the mode for the full data set, just bounce around in the convex hull of the modes for the subsampled datasets without ever hitting the BFGS stopping condition. (Or if it ever stops, it is at one of the subsample modes.)
What is the proposed stopping criterion for this quasi-Newton SGD?

Reply:
The canonical reference is probably this one that includes Byrd and Nocedal as authors:

R. H. Byrd, S. L. Hansen, J. Nocedal, and Y. Singer. A stochastic quasi-Newton method for large-scale optimization. arXiv preprint [arXiv:1401.7020](https://arxiv.org/abs/1401.7020), 2014.  

They take a power-law decay for step size ("steplength" on p. 3, after equation 1.5) which meets the Robbins-Monro conditions. 

In terms of what happens when you leave step size fixed in SGD, it gives you a kind of variational inference where you can think of the iterations as draws from a variational approximation to the posterior.

[A Variational Analysis of Stochastic Gradient Algorithms](https://proceedings.mlr.press/v48/mandt16.html).
Stephan Mandt, Matthew Hoffman, David Blei Proceedings of The 33rd International Conference on Machine Learning, PMLR 48:354-363, 2016.

I've added a bit of discussion of this to the proposal and will cite Mandt et al. as being closest to what you'd get from just plugging into Stan as is.